### ISSUE TITLE
Make `Parser.parse()` return an object

### ISSUE THREAD

**Issue by fwilhe** [jimschubert/kopper#2]
I find it counter intuitive that `Parser.parse()` does not return anything. I think it should return an object of some kind which holds the information that is in the `options` and `_args` fields. Further, I think it would be good if `Parser` objects were immutable.

**Comment by jimschubert** [comment]
These are good points.

In Scala, there's a `Try` which returns a `Success` or `Failure` with data relevant to the result state. For a success, it could contain a match details with information like the Option instance that grabbed the arg, the args and their indexes. What are your thoughts with that approach?

I'll have to look at the immutability of the `Parser` object. I'm sure it's doable, but the typed parser implementation was just easier with a mutable object.

**Comment by fwilhe** [comment]
I did not think about all scenarios, but I would imagine a builder approach is well suited here. I've mocked what I would imagine the api to look like:

```kotlin
package pkg

fun main(args: Array<String>) {
    val options = Builder()
            .name("My Application")
            .description("Does handy things for you")
            .flag("q", listOf("quiet", "silent"), description = "Run silently")
            .option("f", listOf("file"), description = "File name")
            .build().parse()

    if (options.flag("q")) {
        println("Sssshhh!")
    }

    if (options.option("file").endsWith(".md")) {
        println("A markdown fan - I like that")
    }
}
```

This is by no means perfect, it is just what I would do at the moment.

If you're cool with changing the api, I might find some time to do a PR to implement something like this.

**Comment by jimschubert** [comment]
I've made a commit to support method chaining, I'll just need to push it. I didn't implement the full builder pattern because i don't know of that's really beneficial here. I'll push what I have done, and if you'd like to add a Builder I'd be cool with it.

I was looking into a Kotlin equivalent to Scala's Try/Success/Failure. I've implemented these in Kotlin, but I don't think they provide value on top of regular try/catch. 

I like your approach to separating settings for parsing inputs from the parsed options/flags.

I don't mind changing this API at all, because I plan to personally use the typed argument API in the project I've written this for.

**Comment by jimschubert** [comment]
@fwilhe I've opened a PR with an API similar to the one you've provided above. Check it out and let me know what you think. This PR doesn't include a `Builder` as I mentioned above.

**Comment by fwilhe** [comment]
I don't have the time at the moment to take a closer look, but this looks very similar to what I had in mind. I would appreciate a release 0.0.2 with this changes in the near future :)

**Comment by jimschubert** [comment]
Cool. I'll merge and try to get 0.0.2 release done tonight.

**Comment by jimschubert** [comment]
@fwilhe I've published [0.0.2](https://oss.sonatype.org/content/repositories/public/us/jimschubert/kopper/0.0.2/). Check it out!

**Comment by fwilhe** [comment]
Cool, thanks!
